### DISCUSSION TITLE
Disabling authentication for a single page

### DISCUSSION THREAD

**Discussion by firatn** [blitz-js/blitz#3237] [Q&A]
Hi I am working on a new project with Blitz.js. I am relatively new and so far been loving Blitz.js as a product!

I used the Blitz scaffolding to generate pages and queries for a db table called Shipment. The frontend is completed for both

pages/shipments/index.tsx
pages/shipments/[shipmentId].tsx

I wish to use authentication for index.tsx, which is working well. However I am unable to disable authentication for [shipmentId].tsx. The end goal is allowing people to access a specific shipment with a URL link and without logging in.

I tried using `ShipmentsPage.authenticate = false` in [shipmentId].tsx. I am also not sure how to modify the resolver in getShipment.ts to disable authorization.

Any guidance would be highly appreciated. Thank you in advance!

**Comment by noxify** [comment]
Hi,

if your resolver looks like the resolver example from the docs, it should be enough to remove `ctx.session.$authorize()` 
( Docs: https://blitzjs.com/docs/query-resolvers ) 

In case your resolver looks like the example from the "Authorization & Security" docs, it should be enough to remove `resolver.authorize()`
( Docs: https://blitzjs.com/docs/authorization )

For the shipment page, i assume you have something like `Layout.authorize = true` and overwrite the `authorize` on page level ( as explained in the docs ( https://blitzjs.com/docs/authorization#secure-layouts )

If this is the case, your changes should work - But based on the required authorization in the resolver, you will be redirected.
It sounds like, that everything will work after you have removed the authorize check in the resolver.

Please keep in mind, that the api endpoint for the shipment page will be public, too. ( not sure if it's relevant for you )

**Comment by firatn** [comment]
Hi @noxify ,

Thank you for walking me through the details.

I removed `resolver.authorize()` in the resolver in app\shipments\queries\getShipment.ts. I also set the page level `authorize` to false such as `ShowShipmentPage.authenticate = false` in the page [shipmentId].tsx.

Unfortunately the [shipmentId].tsx page does not have a Layout of type BlitzLayout. BlitzLayout seems to be used in pages/shipments/index.tsx but not included in pages/shipments/[shipmentId].tsx. I don't know if this is relevant, but wanted to share this detail.

Kindly note that both getShipments.ts and [shipmentId].tsx are generated by blitz generate.

I am still being redirected to the login page if I enter the URL localhost:3000/shipments/1 as an example. I am unable to find how to disable this redirection in the project. Do you know how to cancel the redirection?

Thank you so much for your time

**Comment by noxify** [comment]
Hi @firatn 

thanks for the feedback.

I tried to reproduce it.

Here my steps to reproduce:

```
# create a new blitz app - TS / npm / formik
blitz new blitz-discussion-3237

# generate the shipment files
blitz generate all shipment

# start the dev server
blitz dev

# open the local env in your browser
# expected result: login page is shown
http://localhost:3000/shipments
```

As next step I changed the following files:

```ts
// app/pages/shipments/index.tsx

/* 
... the auto generated code
*/

// old
// ShipmentsPage.authenticate = true

// new
ShipmentsPage.authenticate = false
ShipmentsPage.getLayout = (page) => <Layout>{page}</Layout>

export default ShipmentsPage
```

```ts
// app/shipments/queries/getShipments.ts

import { paginate, resolver } from "blitz"
import db, { Prisma } from "db"

interface GetShipmentsInput
  extends Pick<Prisma.ShipmentFindManyArgs, "where" | "orderBy" | "skip" | "take"> {}

export default resolver.pipe(
  // disabled the resolver.authorize()
  // resolver.authorize(),
  async ({ where, orderBy, skip = 0, take = 100 }: GetShipmentsInput) => {
    // ...
  }
)
```

The dev server re-compiled everything and after refreshing the `http://localhost:3000/shipments` page in the browser,
I got the expected content from the shipments index page instead of the login page.


( i have tested it also with your mentioned page - the result was the same )

---- 

> Unfortunately the [shipmentId].tsx page does not have a Layout of type BlitzLayout. BlitzLayout seems to be used in pages/shipments/index.tsx but not included in pages/shipments/[shipmentId].tsx. I don't know if this is relevant, but wanted to share this detail.

Unfortunately I can't confirm this - In my generated code had the `BlitzPage`.
Anyway... i removed the parts ( from `ShowShipmentPage: BlitzPage = () ...` to `ShowShipmentPage = () ...` and also the `ShowShipmentPage.getLayout` ) and tried it again.

The result was the same.

Maybe we have a version diff? 

I tested it with the following setup:

```
# Blitz CLI
@blitzjs/cli/0.45.3 darwin-x64 node-v16.13.1

# Used packages in the blitz app
├── @prisma/client@3.11.0
├── @types/preview-email@2.0.1
├── @types/react@17.0.42
├── blitz@0.45.3
├── eslint@7.32.0
├── formik@2.2.9
├── husky@7.0.4
├── lint-staged@11.3.0-beta.2
├── prettier-plugin-prisma@3.11.0
├── prettier@2.6.0
├── pretty-quick@3.1.3
├── preview-email@3.0.5
├── prisma@3.11.0
├── react-dom@18.0.0-alpha-5ca4b0433-20211020
├── react@18.0.0-beta-149b420f6-20211119
├── typescript@4.5.5
└── zod@3.14.2
```

Here the link to the generated blitz app: https://github.com/noxify/blitz-discussion-3237

If this doesn't solve your problem, do you have your app at github or something else available? 
Then we could use this as startpoint to reproduce the issue.

**Comment by firatn** [comment]
Hi @noxify ,

Thank you for walking me through these details, and sharing a sample project on github.

I cloned your project and made the same changes to my project. I can replicate the same results on `blitz-discussion-3237`. However it doesn't work on my project. Our dependencies, including the Blitz versions are all different. I am on Blitz v0.39.0. Let me work on updating all versions on my project to the versions you shared. If the issue still persists, I will be happy to share my project too. I don't want to waste your time if the issue is only a version diff. I will keep you posted.

Thank you so much again for all the support. I really appreciate your time.
